cloud/openstack/os_port.py

```diff
@@ -0,0 +1,391 @@
+    network: foo
+    no_security_groups: True
+
+# Update the existing 'port1' port with multiple security groups (version 1)
+- os_port:
+    state: present
+    auth:
+      auth_url: https://region-b.geo-1.identity.hpcloudsvc.com:35357/v2.0/d
+      username: admin
+      password: admin
+      project_name: admin
+    name: port1
+    security_groups: 1496e8c7-4918-482a-9172-f4f00fc4a3a5,057d4bdf-6d4d-472...
+
+# Update the existing 'port1' port with multiple security groups (version 2)
+- os_port:
+    state: present
+    auth:
+      auth_url: https://region-b.geo-1.identity.hpcloudsvc.com:35357/v2.0/d
+      username: admin
+      password: admin
+      project_name: admin
+    name: port1
+    security_groups:
+      - 1496e8c7-4918-482a-9172-f4f00fc4a3a5
+      - 057d4bdf-6d4d-472...
+'''
+
+RETURN = '''
+id:
+    description: Unique UUID.
+    returned: success
+    type: string
+name:
+    description: Name given to the port.
+    returned: success
+    type: string
+network_id:
+    description: Network ID this port belongs in.
+    returned: success
+    type: string
+security_groups:
+    description: Security group(s) associated with this port.
+    returned: success
+    type: list of strings
+status:
+    description: Port's status.
+    returned: success
+    type: string
+fixed_ips:
+    description: Fixed ip(s) associated with this port.
+    returned: success
+    type: list of dicts
+tenant_id:
+    description: Tenant id associated with this port.
+    returned: success
+    type: string
+allowed_address_pairs:
+    description: Allowed address pairs with this port.
+    returned: success
+    type: list of dicts
+admin_state_up:
+    description: Admin state up flag for this port.
+    returned: success
+    type: bool
+'''
+
+
+def _needs_update(module, port, cloud):
+    """Check for differences in the updatable values.
+
+    NOTE: We don't currently allow name updates.
+    """
+    compare_simple = ['admin_state_up',
+                      'mac_address',
+                      'device_owner',
+                      'device_id']
+    compare_dict = ['allowed_address_pairs',
+                    'extra_dhcp_opt']
+    compare_comma_separated_list = ['security_groups']
+
+    for key in compare_simple:
+        if module.params[key] is not None and module.params[key] != port[key]:
+            return True
+    for key in compare_dict:
+        if module.params[key] is not None and cmp(module.params[key],
+                                                  port[key]) != 0:
+            return True
+    for key in compare_comma_separated_list:
+        if type(module.params[key]) == str:
+                module.params[key] = module.params[key].split(',')
+        if key == 'security_groups':
+            # Retrieve security group id's since security group names are also
+            # supported as valid input.
+            module.params[key] = map(lambda v: cloud.get_security_group(v)['id'], module.params[key])
```

Comment:
Almost there!  :)

get_security_group() will return None if a match is not found. So if a user mistypes a secgroup name, they'll get a weird TypeError exception:

<pre>
TypeError: 'NoneType' object has no attribute '__getitem__'
</pre>

 So, it would probably be best to look for a None return and fail_json() with the bad group.

Also, since we are having to translate group names to IDs twice, maybe we should just do this step up front and pass the UUIDs down to _needs_update() and _compose_port_args()?